PVS-Studio fixes for Renderer{Foundation | Core}

dimhotepus opened the pull request:
Product page: https://www.viva64.com/en/pvs-studio/
How to use for free: https://www.viva64.com/en/b/0600/

dimhotepus committed:
renderercore: Add missed members initialization

dimhotepus committed:
renderercore: Add missed class member initialization

dimhotepus committed:
renderercore: Safe to use macroses (operator priority issues)

dimhotepus committed:
renderercore: Allow extractor to be moved from

m_Extractors[uiIndex] caused const operator[] to
be invoked and result is const auto&.

Const prevented move.

dimhotepus committed:
renderercore: Use ctor initialization list

Do not initilalize class members with default ctor.

dimhotepus commented:
https://github.com/ezEngine/ezEngine/pull/945/commits/9a455d4eb70383e6aadedbf898bf09e3ffc8a5a8

^ This one requires attention. Looks like it fixes regression from 07ae33724330e6c26f3ee1de88eb75ea1ce84524.

C-Core commented on Code/Engine/RendererCore/Meshes/Implementation/MeshBufferResource.cpp:
```diff
@@ -505,6 +505,8 @@ ezResult ezMeshBufferResourceDescriptor::RecomputeNormals()
 
 ezMeshBufferResource::ezMeshBufferResource()
   : ezResource(DoUpdate::OnAnyThread, 1)
+  , m_uiPrimitiveCount(0)
```
Would be even better to initialize in the header directly

jankrassnigg replied:
Indeed for constants that would be nicer. If you have the time to change that, please do, though I would accept the PR this way as well.

jankrassnigg replied:
I wonder whether this is something that we could use Clang-tidy for? @Ingrater ?

dimhotepus replied:
I've initialized class members directly in header.

jankrassnigg commented:
> [9a455d4](https://github.com/ezEngine/ezEngine/commit/9a455d4eb70383e6aadedbf898bf09e3ffc8a5a8)
> 
> ^ This one requires attention. Looks like it fixes regression from [07ae337](https://github.com/ezEngine/ezEngine/commit/07ae33724330e6c26f3ee1de88eb75ea1ce84524).

Thanks for pointing it out. I'll need to think about this carefully. Not sure whether I'll have time for that today.

jankrassnigg commented:
/azp run

jankrassnigg commented on Code/Engine/RendererCore/AnimationSystem/AnimGraph/Implementation/AnimGraphPins.cpp:
```diff
@@ -51,12 +51,8 @@ void ezAnimGraphTriggerOutputPin::SetTriggered(ezAnimGraph& ref_graph, bool bTri
   if (m_iPinIndex < 0)
     return;
 
-  if (!bTriggered)
```
Looking at the rest of the logic, the current code is correct, and this change would break it.
There was a design change at some point, which meant that this code works differently now than originally.

So to get rid of the warning, you can change this line:
const ezInt8 offset = +1; // bTriggered ? +1 : -1;

but keep the if (!bTriggered) return;

It is unclear whether this piece of code will stay that way in the long run, but for now that's good enough.

dimhotepus committed:
renderercore: Make static analyzer happy about already false condition

From  @jankrassnigg:

"Looking at the rest of the logic, the current code is correct, and this
change would break it.  There was a design change at some point,
which meant that this code works differently now than originally.

So to get rid of the warning, you can change this line:
const ezInt8 offset = +1; // bTriggered ? +1 : -1;

but keep the if (!bTriggered) return;

It is unclear whether this piece of code will stay that way in the long
run, but for now that's good enough."
